vendor/github.com/containers/common/libnetwork/pasta/pasta_linux.go

```diff
@@ -26,11 +26,16 @@ import (
 )
 
 const (
-	dnsForwardOpt = "--dns-forward"
+	dnsForwardOpt   = "--dns-forward"
+	mapGuestAddrOpt = "--map-guest-addr"
 
 	// dnsForwardIpv4 static ip used as nameserver address inside the netns,
 	// given this is a "link local" ip it should be very unlikely that it causes conflicts
 	dnsForwardIpv4 = "169.254.0.1"
+
+	// mapGuestAddrIpv4 static ip used as forwarder address inside the netns to reach the host,
+	// given this is a "link local" ip it should be very unlikely that it causes conflicts
+	mapGuestAddrIpv4 = "169.254.0.2"
 )
```

Comment:
So, this is a pre-existing issue with the `--dns-forward` address, but RFC 3927 says the first and last 256 addresses of 169.25.0.0/16 are "reserved for future use".  So it might be wiser to use 169.254.1.XXX.

Reply:
FYI: vendor code is just a copy of the actual upstream code, so it is best to add your comments on https://github.com/containers/common/pull/2136.

Reply:
You are right that there are "reserved":
> The first 256 and last 256 addresses in the 169.254/16
>   prefix are reserved for future use and MUST NOT be selected by a host
>   using this dynamic configuration mechanism.

I am mean sure it is reserved so it might get a special meaning which can break us but using another ip from the proper range is also risking conflicts with real world users
As an example currently all our GCE CI VM's have `nameserver 169.254.169.254` set so it means this ip cannot be reached directly from the container anymore which currently works fine. We do not really know how link local addresses are used in the wild and we should have a default that is unlikely to cause conflicts.

Of course we also might get in trouble if this range will be assigned for some new use. If you have other suggestions of unused ip ranges I am happy to hear them.

Reply:
> FYI: vendor code is just a copy of the actual upstream code, 

I'm aware, but I didn't know how to find the relevant upstream PR..

> so it is best to add your comments on [containers/common#2136](https://github.com/containers/common/pull/2136).

.. until now.

Reply:
> You are right that there are "reserved":
> 
> > The first 256 and last 256 addresses in the 169.254/16
> > prefix are reserved for future use and MUST NOT be selected by a host
> > using this dynamic configuration mechanism.
> 
> I am mean sure it is reserved so it might get a special meaning which can break us but using another ip from the proper range is also risking conflicts with real world users As an example currently all our GCE CI VM's have `nameserver 169.254.169.254` set so it means this ip cannot be reached directly from the container anymore which currently works fine. We do not really know how link local addresses are used in the wild and we should have a default that is unlikely to cause conflicts.

I think it should be ok to use something in the non-reserved range.  AIUI, if guests want to allocate a link-local themselves they should do duplicate address detection via ARP.  Since we know we're the first thing on the link, we should be ok.  Actually, right now pasta responds to all ARPs, so I'm pretty sure a guest *can't* allocate a link local addr (everything will appear to be a duplicate).  Which is something I want to fix, but largely independent of this.

Reply:
It it not that a guest allocates this, if some service on the network is using a link local addresses and we use the same address for dns or map guest option this ip will no longer be reachable from within the container because pasta maps this to a different ip on the host.

To be clear here the GCE instance has `10.128.15.199/32` ip assigned to its interface with
```
default via 10.128.0.1 dev eth0 proto dhcp src 10.128.15.199 metric 100 
10.128.0.1 dev eth0 proto dhcp scope link src 10.128.15.199 metric 100 
```

Yet it uses 169.254.169.254 as nameserver which routes fine in this network. I do not know if other networks do something similar where the link local range is used. So there is very well the chance to break connectivity to that host if we use the same link local ip for the option as that host.

---

And yes I agree this very much "might break" in super weird networks and very likely not a problem for basically everyone. So if others think using 169.254.1.x is better to not use reserved ranges I am fine to change it
@sbrivio-rh @mheon WDYT?

Reply:
Reviewing all this in a bit, and then I'll come back to this.

Reply:
Actually, let me answer here first, as @dgibson already reviewed (I might try to review as well but don't count on it).

> if some service on the network is using a link local addresses and we use the same address for dns or map guest option this ip will no longer be reachable from within the container because pasta maps this to a different ip on the host

I don't see that as a real concern because those addresses are link-local, and anything a container might reasonably expect to access on the same network segment (that's the scope of link-local addresses) is pretty much just DNS nowadays, which we already handle in a special way.

Strictly speaking, I guess Podman should perform duplicate address detection (for IPv4, it's done via ARP, RFC 3927 [2.5](https://datatracker.ietf.org/doc/html/rfc3927#section-2.5)). It's an optimistic form of detection so it doesn't cause additional delays in a general case, but it's not exceedingly simple either, and the risk looks so low to me that I'm not sure it makes sense to implement this.

So, yeah, I would use 168.254.1.x/24.

The risk of using reserved ranges is not just that they might be assigned to something one day, but you might fail sanity checks in applications, or in a future version of the kernel, etc.

Reply:
This kind of all comes back to what we consider "the link".  Is it just the virtual link from the container to pasta, or does it include things that are local on one of the host's links.  At the moment we're not really consistently either, which is something I have slowly progressing plans to sanitize.

Allowing access to things on the host link is arguably "more transparent", and does let you reach those sites.  However it makes it much less clear what the semantics should be in the case of multiple host links, or a disconnected host, or a host moving from one link to another.  Plus, I'm not sure it's even possible to adequately maintain the illusion of being on the host link in enough cases.

So, I'm generally trying to push us in the direction of considering the pasta<->container connection "the link", at least in default configurations.  This has the extremely convenient additional effect that it means that link-local addresses all become free to allocate for internal purposes.

Reply:
updated in https://github.com/containers/common/pull/2136